### PR TITLE
tcp proxy: enable to specify a path for HTTP tunneling with POST method

### LINKED ISSUE #24038: Path cannot be overwritten when using TCP tunneling
*Title*: Path is hardcoded as "/" and cannot be overwritten when using TCP tunneling

*Description*: When using HTTP upgrades for TCP tunneling it is sometimes required to select a custom path, for HTTP POST request, for instance. Currently path is hard coded as "/" in the source code. Feature suggested is adding an option to overwrite the path by configuration.

[optional *Relevant Links*:]
Slack discussion over this topic: https://envoyproxy.slack.com/archives/C78HA81DH/p1666598752138609

### PR DESCRIPTION
**Pull request by soulxu**
Signed-off-by: He Jie Xu <hejie.xu@intel.com>

Commit Message: tcp proxy: enable to specify a path for HTTP tunneling with POST method
Additional Description:
Risk Level: low
Testing: unittest
Docs Changes: API doc
Release Notes: new feature
Fixes #24038

### CODE REVIEW

**Comment by soulxu** [comment]
cc @alyssawilk

File: source/common/tcp_proxy/upstream.cc [hunk 1]
```diff
@@ -287,10 +287,10 @@ void Http2Upstream::setRequestEncoder(Http::RequestEncoder& request_encoder, boo
   });
 
   if (config_.usePost()) {
-    headers->addReference(Http::Headers::get().Path, "/");
+    headers->addReference(Http::Headers::get().Path, config_.postPath());
     headers->addReference(Http::Headers::get().Scheme, scheme);
   } else if (!Runtime::runtimeFeatureEnabled("envoy.reloadable_features.use_rfc_connect")) {
-    headers->addReference(Http::Headers::get().Path, "/");
+    headers->addReference(Http::Headers::get().Path, config_.postPath());
```
Comment (lizan) [inline source/common/tcp_proxy/upstream.cc:293]: I don't think you should set path for CONNECT?

File: api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto [hunk 2]
```diff
@@ -103,6 +104,11 @@ message TcpProxy {
     // Save the response headers to the downstream info filter state for consumption
     // by the network filters. The filter state key is ``envoy.tcp_proxy.propagate_response_headers``.
     bool propagate_response_headers = 4;
+
+    // The path used with POST method. Default path is ``\``. If post path is specified and
```
Comment (lizan) [inline api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto:108]: default is `/`.

File: source/common/tcp_proxy/upstream.cc (hunk 1 above)
Comment (kyessenov) [reply source/common/tcp_proxy/upstream.cc:293]: It needs to force extended CONNECT https://www.rfc-editor.org/rfc/rfc8441, i.e. set :protocol.

Comment (soulxu) [reply source/common/tcp_proxy/upstream.cc:293]: got it, I can allow the path when use_rfc_connect is true.

File: api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto (hunk 2 above)
Comment (soulxu) [reply api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto:108]: oops, got it, thanks!

File: source/common/tcp_proxy/upstream.cc (hunk 1 above)
Comment (soulxu) [reply source/common/tcp_proxy/upstream.cc:293]: emm... this is deprecated, we said this is deprecated extended CONNECT rfc, I'm confused https://github.com/envoyproxy/envoy/blob/b3456877cdcf418294b2b3b3db21487e0c425059/changelogs/1.23.0.yaml#L152

Comment (soulxu) [reply source/common/tcp_proxy/upstream.cc:293]: There is a long conversation about this https://github.com/envoyproxy/envoy/issues/20378, I think it means this is a wrong usage of extended CONNECT. @kyessenov, so we needn't support custom path for it, right?

Comment (kyessenov) [reply source/common/tcp_proxy/upstream.cc:293]: No, I just mean that setting path must be coupled with extended connect. You shouldn't set path for regular connect.

Comment (soulxu) [reply source/common/tcp_proxy/upstream.cc:293]: oh, got it, will update, thanks

**Comment by jmarantz** [comment]
@lizan can you take another look?

**Review by ssripadham — APPROVED** [review]

**Review by ohadvano — APPROVED** [review]

**Review by lizan — APPROVED** [review]

**Pull request merged**